Add ProtoKeeb keyboard request

atechofficials opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

The Protokeeb is a versatile 16-key (4 x 4) Numpad/Macropad/MIDI device designed by A-Tech Officials. It is perfect for anyone looking to get started with custom keyboards. This open-source hardware project is based on a Raspberry Pi Pico development board, making it easy to learn, build, and modify. Designed for developers and enthusiasts alike, the Protokeeb features per-key south-facing ARGB LEDs and much more.

For more information, visit the [Protokeeb GitHub page](https://github.com/atechofficials/protokeeb).

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

waffle87 commented on keyboards/protokeeb/keymaps/via/config.h:
This file should be removed.

atechofficials replied:
I have removed the "via" keymap's config.h file as requested

waffle87 commented on keyboards/protokeeb/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,75 @@
+ */
+
+#include QMK_KEYBOARD_H
+#include "quantum.h"
```
```suggestion
```

waffle87 commented on keyboards/protokeeb/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,75 @@
+#include "quantum.h"
+
+enum protokeeb_keymap_layers {
+    LAYER_BASE = 0,
+    LAYER_LOWER = 1,
+    DYNAMIC_LAYER1 = 2,
+    DYNAMIC_LAYER2 = 3
```
```suggestion
    LAYER_BASE
    LAYER_LOWER
```

waffle87 commented on keyboards/protokeeb/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,75 @@
+	  RGB_HUI, RGB_HUD, RGB_SAI, RGB_SAD,
+	  RGB_VAI, RGB_VAD, RGB_SPI, RGB_SPD,
+	  KC_TRNS, RGB_MODE_SWIRL, QK_DEBUG_TOGGLE, QK_BOOT
+	),
+
+  [DYNAMIC_LAYER1] = LAYOUT_ortho_4x4( /* Empty for dynamic keymaps */
+	  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+	  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+	  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+	  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS
+	),
+
+  [DYNAMIC_LAYER2] = LAYOUT_ortho_4x4( /* Empty for dynamic keymaps */
+	  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+	  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+	  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+	  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS
+	),
```
```suggestion
	)
```

waffle87 commented on keyboards/protokeeb/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,75 @@
+#if defined(ENCODER_MAP_ENABLE)
+const uint16_t PROGMEM encoder_map[][NUM_ENCODERS][NUM_DIRECTIONS] = {
+    [LAYER_BASE] = { ENCODER_CCW_CW(KC_VOLD, KC_VOLU) },
+    [LAYER_LOWER] = { ENCODER_CCW_CW(KC_BRIGHTNESS_DOWN, KC_BRIGHTNESS_UP) },
+		[DYNAMIC_LAYER1] = { ENCODER_CCW_CW(KC_TRNS, KC_TRNS) },
+    [DYNAMIC_LAYER2] = { ENCODER_CCW_CW(KC_TRNS, KC_TRNS) },
```
```suggestion
    [LAYER_LOWER] = { ENCODER_CCW_CW(KC_BRIGHTNESS_DOWN, KC_BRIGHTNESS_UP) }
```

waffle87 commented on keyboards/protokeeb/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,75 @@
+    [DYNAMIC_LAYER2] = { ENCODER_CCW_CW(KC_TRNS, KC_TRNS) },
+};
+#endif
+
+layer_state_t layer_state_set_user(layer_state_t state) {
+    return update_tri_layer_state(state, LAYER_LOWER, DYNAMIC_LAYER1, DYNAMIC_LAYER2);
+}
```
```suggestion
```

waffle87 commented on keyboards/protokeeb/keymaps/via/rules.mk:
```diff
@@ -0,0 +1,13 @@
+
+# DIP Switch Keymap support
+DIP_SWITCH_MAP_ENABLE = no
+
+TRI_LAYER_ENABLE = yes
```
This is conditionally enabled with `VIA_ENABLE`.
```suggestion
```

waffle87 commented on keyboards/protokeeb/rev1/config.h:
```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+/* ARGB LED Byte order */
+#undef WS2812_BYTE_ORDER
```
```suggestion
```

waffle87 commented on keyboards/protokeeb/rev1/halconf.h:
This file should be removed.

atechofficials replied:
I have removed the halconf.h file as requested.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+{
+    "usb": {
+        "pid": "0x1717",
+        "device_version": "1.0.0",
+        "force_nkro": false,
+        "max_power": 500,
+        "no_startup_check": false,
+        "polling_interval": 1,
+        "shared_endpoint": {
+            "keyboard": true,
+            "mouse": false
+        },
+        "suspend_wakeup_delay": 0,
+        "wait_for_enumeration": false
```
These are all the default behaviours.
```suggestion
        "device_version": "1.0.0"
```

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+    "board": "GENERIC_RP_RP2040",
+    "bootloader": "rp2040",
+    "processor": "RP2040",
+    "build": {
+        "debounce_type": "sym_defer_g",
+        "firmware_format": "uf2",
+        "lto": false
+    },
```
As above, all the defaults.
```suggestion
```

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+        "extrakey": true,
+        "nkro": true
+    },
+    "qmk": {
+        "locking": {
+            "enabled": false,
+            "resync": false
+        },
+        "tap_capslock_delay": 80,
+        "tap_keycode_delay": 0
+    },
```
And here.
```suggestion
```

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
Please run `qmk format-json` on this file.

atechofficials replied:
Done.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+                    "matrix": [2, 3],
+                    "x": 3,
+                    "y": 2,
+                    "w": 1,
+                    "h": 1
+                },
+                {
+                    "label": "0",
+                    "matrix": [3, 0],
+                    "x": 0,
+                    "y": 3,
+                    "w": 1,
+                    "h": 1
+                },
+                {
+                    "label": ".",
+                    "matrix": [3, 1],
+                    "x": 1,
+                    "y": 3,
+                    "w": 1,
+                    "h": 1
+                },
+                {
+                    "label": "ENTER",
+                    "matrix": [3, 2],
+                    "x": 2,
+                    "y": 3,
+                    "w": 1,
+                    "h": 1
+                },
+                {
+                    "label": "+",
+                    "matrix": [3, 3],
+                    "x": 3,
+                    "y": 3,
+                    "w": 1,
+                    "h": 1
+                }
+            ]
+        },
```
Given that these layouts are all functionally identical, only one should be included.
```suggestion
```

atechofficials replied:
Ok. I have removed the "LAYOUT" layout from the "keyboard.json" file as requested. Also, I have updated the default keymap's "keymap.c" file as the "LAYOUT" was being used in that file.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+                    "matrix": [2, 3],
+                    "x": 3,
+                    "y": 2,
+                    "w": 1,
+                    "h": 1
+                },
+                {
+                    "label": "MO(3)",
+                    "matrix": [3, 0],
+                    "x": 0,
+                    "y": 3,
+                    "w": 1,
+                    "h": 1
+                },
+                {
+                    "label": "TG(1)",
+                    "matrix": [3, 1],
+                    "x": 1,
+                    "y": 3,
+                    "w": 1,
+                    "h": 1
+                },
+                {
+                    "label": "TG(2)",
+                    "matrix": [3, 2],
+                    "x": 2,
+                    "y": 3,
+                    "w": 1,
+                    "h": 1
+                },
+                {
+                    "label": "MI_TOGG",
+                    "matrix": [3, 3],
+                    "x": 3,
+                    "y": 3,
+                    "w": 1,
+                    "h": 1
+                }
+            ]
+        }
```
```suggestion
        }
```

atechofficials replied:
Ok. I have removed the "LAYOUT_midi" layout from the "keyboard.json" file as requested. Also, this layout was not identical to any of the existing layouts in "keyboard.json" file.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+                    "label": "7",
+                    "matrix": [0, 0],
+                    "x": 0,
+                    "y": 0,
+                    "w": 1,
+                    "h": 1
```
```suggestion
                    "y": 0
```

atechofficials replied:
Done.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+                    "label": "8",
+                    "matrix": [0, 1],
+                    "x": 1,
+                    "y": 0,
+                    "w": 1,
+                    "h": 1
```
```suggestion
                    "y": 0
```

atechofficials replied:
Done.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+                    "label": "9",
+                    "matrix": [0, 2],
+                    "x": 2,
+                    "y": 0,
+                    "w": 1,
+                    "h": 1
```
```suggestion
                    "y": 0
```

atechofficials replied:
Done.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+                    "label": "/",
+                    "matrix": [0, 3],
+                    "x": 3,
+                    "y": 0,
+                    "w": 1,
+                    "h": 1
```
```suggestion
                    "y": 0
```

atechofficials replied:
Done.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+                    "label": "4",
+                    "matrix": [1, 0],
+                    "x": 0,
+                    "y": 1,
+                    "w": 1,
+                    "h": 1
```
```suggestion
                    "y": 1
```

atechofficials replied:
Done.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+                    "label": "5",
+                    "matrix": [1, 1],
+                    "x": 1,
+                    "y": 1,
+                    "w": 1,
+                    "h": 1
```
```suggestion
                    "y": 1
```

atechofficials replied:
Done.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+                    "label": "6",
+                    "matrix": [1, 2],
+                    "x": 2,
+                    "y": 1,
+                    "w": 1,
+                    "h": 1
```
```suggestion
                    "y": 1
```

atechofficials replied:
Done.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+                    "label": "*",
+                    "matrix": [1, 3],
+                    "x": 3,
+                    "y": 1,
+                    "w": 1,
+                    "h": 1
```
```suggestion
                    "y": 1
```

atechofficials replied:
Done.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+                    "label": "1",
+                    "matrix": [2, 0],
+                    "x": 0,
+                    "y": 2,
+                    "w": 1,
+                    "h": 1
```
```suggestion
                    "y": 2
```

atechofficials replied:
Done.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+                    "label": "2",
+                    "matrix": [2, 1],
+                    "x": 1,
+                    "y": 2,
+                    "w": 1,
+                    "h": 1
```
```suggestion
                    "y": 2
```

atechofficials replied:
Done.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+                    "label": "3",
+                    "matrix": [2, 2],
+                    "x": 2,
+                    "y": 2,
+                    "w": 1,
+                    "h": 1
```
```suggestion
                    "y": 2
```

atechofficials replied:
Done.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+                    "label": "-",
+                    "matrix": [2, 3],
+                    "x": 3,
+                    "y": 2,
+                    "w": 1,
+                    "h": 1
```
```suggestion
                    "y": 2
```

atechofficials replied:
Done.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+                    "label": "0",
+                    "matrix": [3, 0],
+                    "x": 0,
+                    "y": 3,
+                    "w": 1,
+                    "h": 1
```
```suggestion
                    "y": 3
```

atechofficials replied:
Done.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+                    "label": ".",
+                    "matrix": [3, 1],
+                    "x": 1,
+                    "y": 3,
+                    "w": 1,
+                    "h": 1
```
```suggestion
                    "y": 3
```

atechofficials replied:
Done.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+                    "label": "ENTER",
+                    "matrix": [3, 2],
+                    "x": 2,
+                    "y": 3,
+                    "w": 1,
+                    "h": 1
```
```suggestion
                    "y": 3
```

atechofficials replied:
Done.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+                    "label": "+",
+                    "matrix": [3, 3],
+                    "x": 3,
+                    "y": 3,
+                    "w": 1,
+                    "h": 1
```
```suggestion
                    "y": 3
```

atechofficials replied:
Done.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+            ]
+        }
+    },
+    "audio": {
+        "default": {
+            "on": false,
+            "clicky": false
+        },
+        "macro_beep": false,
+        "pins": ["GP22"],
+        "voices": false
+    },
```
```suggestion
    "audio": {
        "default": {
            "on": false
        },
        "pins": ["GP22"],
    },
```

atechofficials replied:
Done. Also, the comma ( , ) needs to be removed after the "pins": ["GP22"] line for a valid error-free JSON file.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+        "pins": ["GP22"],
+        "voices": false
+    },
+    "bootmagic": {
+        "enabled": true,
+        "matrix": [0, 0]
+    },
```
```suggestion
```

atechofficials replied:
Done.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+        "matrix": [0, 0]
+    },
+    "dip_switch": {
+        "enabled": true,
```
Enable at the keyboard level but disabled at all the keymap levels?

atechofficials replied:
I am using the dip switch functionality to use the encoder's built-in push button. Since the push button is not connected to the switch matrix instead it is directly connected to the GPIO pin of the RP2040 MCU. So, to use it on this keyboard I have to use the dip switch functionality to assign a fixed keycode to the push button. The user can use the push button to send a fixed keycode (KC_MUTE in this case) no matter what keymap they use with the keyboard.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+            "speed": 0
+        },
+        "driver": "ws2812",
+        "layers": {
+            "blink": false,
+            "enabled": false,
+            "max": 4
+        },
```
```suggestion
```

atechofficials replied:
Done.

waffle87 commented on keyboards/protokeeb/rev1/mcuconf.h:
```diff
@@ -0,0 +1,36 @@
+#undef RP_I2C_USE_I2C1
+#define RP_I2C_USE_I2C1 FALSE
+
+// #define RP_I2C_BUSY_TIMEOUT 50
+// #define RP_I2C_USE_DMA TRUE
+
```
```suggestion
```

waffle87 commented on keyboards/protokeeb/rev1/mcuconf.h:
```diff
@@ -0,0 +1,36 @@
+// #define RP_I2C_BUSY_TIMEOUT 50
+// #define RP_I2C_USE_DMA TRUE
+
+/* RP2040 SPI Driver Selection */
+#undef RP_SPI_USE_SPI0
+#define RP_SPI_USE_SPI0 TRUE
+
+#undef RP_SPI_USE_SPI1
+#define RP_SPI_USE_SPI1 FALSE
```
What is SPI used for on this keyboard? If nothing, this should be removed.

atechofficials replied:
I was testing other functionalities of the QMK firmware. Removed SPI preprocessor directives.

waffle87 commented on keyboards/protokeeb/rev1/readme.md:
````diff
@@ -0,0 +1,44 @@
+
+Make example for this keyboard (after setting up your build environment):
+
+```go
+make protokeeb/rev1:default
+```
+
+Flashing example for this keyboard:
+
+```go
+make protokeeb/rev1:default:flash
+```
````
```suggestion
    make protokeeb/rev1:default

Flashing example for this keyboard:

    make protokeeb/rev1:default:flash

```

waffle87 commented on keyboards/protokeeb/rev1/rev1.c:
```diff
@@ -0,0 +1,40 @@
+							// Toggle Host Mute
+							tap_code(KC_MUTE);
+						}
+						// Encoder Push Button Released
+						else {
+							// Do Nothing
+						}
```
```suggestion
```

waffle87 commented on keyboards/protokeeb/rev1/rev1.c:
Please run `qmk format-c` on this file.

atechofficials replied:
Done.

waffle87 commented on keyboards/protokeeb/rev1/rev1.h:
This file should be removed.

atechofficials replied:
Done.

waffle87 commented on keyboards/protokeeb/rev1/rev1.c:
```diff
@@ -0,0 +1,40 @@
+ * If not, see <https://www.gnu.org/licenses/gpl-3.0.html>.
+ */
+
+#include "rev1.h"
```
```suggestion
#include "quantum.h"
```

waffle87 commented on keyboards/protokeeb/rev1/rules.mk:
This file should be removed. Since OS detection is not actually utilised anywhere, it should be enabled. Encoders should be enable in `keyboard.json`.

atechofficials replied:
Done.

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,507 @@
+        "bootmagic": true,
+        "rgblight": true,
+        "extrakey": true,
+        "nkro": true
```
```suggestion
        "nkro": true,
        "encoder": true
```

atechofficials commented on keyboards/protokeeb/keymaps/debug/keymap.c:
I have removed the debug keymap directory as requested.

atechofficials commented on keyboards/protokeeb/keymaps/default_ortho_4x4/keymap.c:
This keymap was created to showcase QMK's tri-layer functionality to end users. Its purpose was to offer users a customizable multi-layer option for their Protokeeb keyboard. However, if this does not align with the guidelines of the QMK firmware, I will remove it.

atechofficials replied:
Please confirm the same. This keymap removal is necessary or not?

waffle87 replied:
I don't see why the `default` keymap couldn't be used to showcase the tri layer functionality?

atechofficials replied:
According to the QMK [PR Checklist](https://docs.qmk.fm/pr_checklist#keyboard-prs) "the default (and via) keymaps should be pristine and bare minimum to be used as a clean slate for another user to develop their own user-specific keymap". This was the reason I created a separate keymap with the tri-layer functionality.

waffle87 replied:
Tri layer functionality would be something acceptable in a `default` keymap. That comment more so concerns extravagant and confusing feature implementations that aren't necessary for the board/most people won't even use.

atechofficials replied:
Ok, understood. I am going to implement the tri-layer functionality in the default keymap in future updates.

atechofficials commented:
All the required changes to the keyboard files are done. Please review the changes and further process the pull request.

drashna reviewed (commented):
Wait, so the encoder switch pins are hooked up outside of the matrix?

atechofficials commented:
> Wait, so the encoder switch pins are hooked up outside of the matrix?

Yes, the encoder push button is not a part of the switch matrix. It is directly connected to the MCU pin. Here is the link to the [Protokeeb rev1 schematics](https://github.com/atechofficials/protokeeb/blob/master/PCB/Rev1.0/AT-ProtoKeeb_rev1/ProtoKeeb_rev1_Schematics.pdf) for your reference. Also, check the Protokeeb keyboard's [rev1.c](https://github.com/atechofficials/qmk_firmware/blob/dev_branch/keyboards/protokeeb/rev1/rev1.c) file to understand the encoder push button functionality in this keyboard.

drashna commented:
Is there a specific reason that it wasn't included, or was that just a mistake and this is correction for it after the fact? 

If it's correction after the fact, then using the dipswitch code really isn't a good way to do this.  

You could add an additional column, and add a local `matrix_read_cols_on_row` that reads from the specific pin for the encoder at the column/row spot.   It would allow the encoder to act like it was hooked into the switch matrix, normally.  However, it would no longer be an ortho 4x4 layout.  But it would allow for proper via support for the encoder, for instance.

atechofficials commented:
> Is there a specific reason that it wasn't included, or was that just a mistake and this is correction for it after the fact?
> 
> If it's correction after the fact, then using the dipswitch code really isn't a good way to do this.
> 
> You could add an additional column, and add a local `matrix_read_cols_on_row` that reads from the specific pin for the encoder at the column/row spot. It would allow the encoder to act like it was hooked into the switch matrix, normally. However, it would no longer be an ortho 4x4 layout. But it would allow for proper via support for the encoder, for instance.

Since it is my first custom-designed mechanical keyboard, it has some flaws which I will rectify in the upcoming revisions. To make the encoder push button functional instead of not using it at all because of the way it is electrically connected, I had to figure out some way to make it functional and upon searching the QMK documentation I ended up selecting the dipswitch functionality of the QMK firmware. 
Thank you for suggesting a better way to implement the encoder push button functionality, instead of using the dipswitch functionality. I would try this approach and test if it works as expected.

atechofficials commented:
## Project Update:
- Successfully implemented the suggested method for the encoder push button functionality. 
- Firmware compiled successfully and also tested the encoder push button and it is working as intended. VIA keymap is also working as expected. 

The only problem that I have found with [VIA configurator](https://usevia.app/) is that there is no method to implement/display the push button key and the encoder as a single entity in the VIA configurator UI. Also, to let the end users access the push button key to assign their keycodes I also modified the Protokeeb Rev1 VIA Layout JSON file which I am going to upload to the [VIA GitHub repository](https://github.com/the-via/keyboards) by creating a pull request later.

## ProtoKeeb Rev1 VIA JSON file

```JSON
{
    "name": "Protokeeb",
    "vendorId": "0x419A",
    "productId": "0x1717",
    "matrix": {
        "rows": 4,
        "cols": 5
    },
    "keycodes": ["qmk_lighting"],
    "menus": ["qmk_rgblight"],
    "layouts": {
        "keymap": [
            [
                {
                    "y": 0.3,
                    "x": 2,
                    "a": 4
                },
                "0,4",
                {
                    "a": 7
                },
                "e0"
            ],
            [
                {
                    "a": 4
                },
                "0,0",
                "0,1",
                "0,2",
                "0,3"
            ],
            [
                "1,0",
                "1,1",
                "1,2",
                "1,3"
            ],
            [
                "2,0",
                "2,1",
                "2,2",
                "2,3"
            ],
            [
                {
                    "c": "#aaaaaa"
                },
                "3,0",
                {
                    "c": "#cccccc"
                },
                "3,1",
                "3,2",
                "3,3"
            ]
        ]
    }
}
```

waffle87 commented on keyboards/protokeeb/rev1/rev1.c:
Please run `qmk format-c` on this file. (Good general rule of thumb before committing)

atechofficials replied:
Done.

waffle87 commented on keyboards/protokeeb/rules.mk:
```diff
@@ -0,0 +1,5 @@
+# Build Options
+# Change yes to no to diable feature
+
```
```suggestion
```

waffle87 commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,88 @@
+        "pin": "GP10"
+    },
+    "layouts": {
+        "LAYOUT_custom_matrix": {
```
And update the keymaps accordingly as well. 
```suggestion
        "LAYOUT": {
```

atechofficials replied:
I have updated all the keymap files according to the layout's name in the keyboard.json file.

waffle87 commented on keyboards/protokeeb/keymaps/default/rules.mk:
```diff
@@ -0,0 +1,8 @@
+
+# Rotary Encoder Keymap support
+ENCODER_MAP_ENABLE = yes
+
+# DIP Switch Keymap support
+DIP_SWITCH_MAP_ENABLE = no
```
```suggestion
```

waffle87 commented on keyboards/protokeeb/keymaps/midi/rules.mk:
```diff
@@ -0,0 +1,11 @@
+
+# Rotary Encoder Keymap support
+ENCODER_MAP_ENABLE = yes
+
+# DIP Switch Keymap support
+DIP_SWITCH_MAP_ENABLE = no
```
```suggestion
```

waffle87 commented on keyboards/protokeeb/keymaps/via/rules.mk:
```diff
@@ -0,0 +1,11 @@
+
+# Rotary Encoder Keymap support
+ENCODER_MAP_ENABLE = yes
+
+# DIP Switch Keymap support
+DIP_SWITCH_MAP_ENABLE = no
```
```suggestion
```

atechofficials commented:
After merging the master branch of my QMK fork with the "dev_branch" to keep it updated with the latest QMK firmware changes, the QMK CI Build check in this pull request is taking significantly longer than my previous commits. It's been 11 hours and the test is still not completed. Can you please look into this issue?

atechofficials commented:
Just a heads up that all the requested changes have been made in the Protokeeb keyboard firmware files. The firmware has compiled successfully on my local machine, and I've tested everything to ensure it works as intended. Please take a moment to review the changes and then proceed with the pull request.

fauxpark commented:
You've merged incorrectly, reverting a bunch of newer changes upstream. Please fix this.

atechofficials commented:
> You've merged incorrectly, reverting a bunch of newer changes upstream. Please fix this.

I've merged the master branch of QMK with my QMK fork's dev_branch to address this issue. If there are any additional steps or suggestions regarding this matter, please feel free to let me know.

atechofficials commented:
Dear QMK Maintainers,
@fauxpark @drashna @waffle87 

I wanted to follow up on my previous communication regarding the pull request I submitted. Despite having addressed all the requested changes, I have not yet received a response from any of the QMK Maintainers. I understand that you receive numerous pull requests daily and that reviewing and responding to each one promptly may not be feasible. I kindly request your attention to further process my pull request at your earliest convenience.

Thank you for your time and consideration.

fauxpark commented:
Please look at the "Files changed" tab. There are a bunch of irrelevant changes that need to be undone.

atechofficials commented:
> Please look at the "Files changed" tab. There are a bunch of irrelevant changes that need to be undone.

My knowledge of git is limited. Please guide me through the steps needed to fix the issue.

atechofficials commented:
Fixed the issues that occurred due to the incorrect merging of branches and content. Reverted back to the commit where everything was working as intended.

atechofficials commented:
Dear QMK Maintainers @waffle87 @drashna

Subject: Review Request for Pull Request

I have completed all the requested changes. The firmware has been successfully compiled and everything is working as intended. I submitted a request for a re-review of my pull request 3 days ago. I kindly request the QMK maintainers to review my pull request as soon as possible.

drashna commented on keyboards/protokeeb/rev1/config.h:
```diff
@@ -0,0 +1,55 @@
+/* LED to blink when entering bootloader mode */
+#define RP2040_BOOTLOADER_DOUBLE_TAP_RESET_LED GP25
+
+#if defined (ENCODER_ENABLE)
+#define ENCODER_MAP_KEY_DELAY 10
+#endif
```
You don't need the ifdef around this.  If encoder map isn't enabled, this doesn't do anything, so it's safe to keep.
```suggestion
#define ENCODER_MAP_KEY_DELAY 10
```

atechofficials replied:
Done.

drashna commented on keyboards/protokeeb/rev1/config.h:
```diff
@@ -0,0 +1,55 @@
+#define RGBLIGHT_EFFECT_SWIRL_RANGE 255
+#define RGBLIGHT_EFFECT_SNAKE_LENGTH 4
+#define RGBLIGHT_EFFECT_TWINKLE_LIFE 200
+#define RGBLIGHT_EFFECT_TWINKLE_PROBABILITY 1/127
```
Good practice to  wrap stuff like this with parenthesis.
```suggestion
#define RGBLIGHT_EFFECT_TWINKLE_PROBABILITY (1/127)
```

atechofficials replied:
Done.

atechofficials commented:
Hello @drashna and @waffle87,

It's been two weeks since I submitted this pull request. I've made all the requested file modifications and optimizations from the QMK maintainers. I'm not sure why there's a delay in processing this pull request. Could you please either merge it or let me know if my keyboard contribution isn't considered worthy for the QMK project?

Thank you!

fauxpark commented on keyboards/protokeeb/rev1/rev1.c:
```diff
@@ -0,0 +1,28 @@
+// Override the matrix read function to include the encoder button
+void matrix_read_cols_on_row(matrix_row_t current_matrix[], uint8_t current_row) {
+    if (current_row == ENCODER_BUTTON_ROW) {
+        if (readPin(ENCODER_BUTTON_PIN) == 0) {
```
```suggestion
        if (gpio_read_pin(ENCODER_BUTTON_PIN) == 0) {
```

atechofficials replied:
Done.

fauxpark commented on keyboards/protokeeb/rev1/keyboard.json:
Please run `qmk format-json -i` on this file.

atechofficials replied:
Done.

fauxpark commented on keyboards/protokeeb/keymaps/via/rules.mk:
```diff
@@ -0,0 +1,8 @@
+# Build Options Protokeeb rev1
+# Change yes to no to disable features
+
+# Enable Via support
+VIA_ENABLE = yes
+
+# Rotary Encoder Keymap support
+ENCODER_MAP_ENABLE = yes
```
```suggestion
VIA_ENABLE = yes
ENCODER_MAP_ENABLE = yes
```

atechofficials replied:
Done.

fauxpark commented on keyboards/protokeeb/rules.mk:
This file can be deleted.

atechofficials replied:
Done.

fauxpark commented on keyboards/protokeeb/keymaps/midi/rules.mk:
```diff
@@ -0,0 +1,8 @@
+# Build Options
+# Change yes to no to disable feature
+
+# Enable MIDI support
+MIDI_ENABLE = yes
+
+# Rotary Encoder Keymap support
+ENCODER_MAP_ENABLE = yes
```
```suggestion
MIDI_ENABLE = yes
ENCODER_MAP_ENABLE = yes
```

atechofficials replied:
Done.

fauxpark commented on keyboards/protokeeb/keymaps/default/rules.mk:
```diff
@@ -0,0 +1,8 @@
+# Build Options Protokeeb rev1
+# Change yes to no to disable features
+
+# Rotary Encoder Keymap support
+ENCODER_MAP_ENABLE = yes
+
+# Enable Tri Layer support
+TRI_LAYER_ENABLE = yes
```
```suggestion
ENCODER_MAP_ENABLE = yes
TRI_LAYER_ENABLE = yes
```

atechofficials replied:
Done.

atechofficials commented:
Hello @fauxpark,
All the requested changes are done, please further process this pull request.

drashna commented on keyboards/protokeeb/rev1/config.h:
Did you plan on additional revisions, here?

atechofficials replied:
Yes, I am going to design additional revisions for the Protokeeb keyboard. As mentioned earlier, rev1 has some minor flaws, which will be rectified in the upcoming revision.

fauxpark commented on keyboards/protokeeb/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,58 @@
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    [LAYER_BASE] = LAYOUT(/* Base: Numpad without Num Lock*/
+                          KC_KP_7, KC_KP_8, KC_KP_9, KC_KP_SLASH, KC_MUTE, KC_KP_4, KC_KP_5, KC_KP_6, KC_KP_ASTERISK, KC_KP_1, KC_KP_2, KC_KP_3, KC_KP_MINUS, LOWER, KC_KP_DOT, RAISE, KC_KP_PLUS),
+
+    [LAYER_LOWER] = LAYOUT(/* RGB Control */
+                           RGB_MODE_PLAIN, RGB_MODE_FORWARD, RGB_MODE_REVERSE, QK_DEBUG_TOGGLE, RGB_TOG, RGB_HUI, RGB_HUD, RGB_SAI, RGB_SAD, RGB_VAI, RGB_VAD, RGB_SPI, RGB_SPD, KC_TRNS, RGB_MODE_SWIRL, KC_TRNS, QK_BOOT),
+
+    [LAYER_RAISE] = LAYOUT(/* Media Control and Quick Launch*/
+                           KC_MEDIA_PLAY_PAUSE, KC_MEDIA_PREV_TRACK, KC_MEDIA_NEXT_TRACK, KC_MEDIA_STOP, KC_MUTE, KC_CALCULATOR, KC_MY_COMPUTER, KC_CONTROL_PANEL, KC_SYSTEM_POWER, KC_PRINT_SCREEN, KC_HOME, KC_INSERT, KC_DELETE, KC_TRNS, KC_NO, KC_TRNS, QK_REBOOT),
+
+    [LAYER_ADJUST] = LAYOUT(/* Function Keys */
+                            KC_F1, KC_F2, KC_F3, KC_F4, KC_MUTE, KC_F5, KC_F6, KC_F7, KC_F8, KC_F9, KC_F10, KC_F11, KC_F12, KC_TRNS, KC_NO, KC_TRNS, KC_NO),
+};
+
+#if defined(ENCODER_MAP_ENABLE)
+const uint16_t PROGMEM encoder_map[][NUM_ENCODERS][NUM_DIRECTIONS] = {
+    [LAYER_BASE]   = {ENCODER_CCW_CW(KC_VOLD, KC_VOLU)},
+    [LAYER_LOWER]  = {ENCODER_CCW_CW(KC_BRIGHTNESS_DOWN, KC_BRIGHTNESS_UP)},
+    [LAYER_RAISE]  = {ENCODER_CCW_CW(KC_VOLD, KC_VOLU)},
+    [LAYER_ADJUST] = {ENCODER_CCW_CW(KC_TRNS, KC_TRNS)},
```
Prefer the short keycode names in keymaps.

atechofficials replied:
Thank you for sharing your input. I will take this into consideration for my future development efforts.

atechofficials commented:
Please provide a comprehensive list of suggested changes and optimizations for the firmware files of this keyboard. The current process, which involves three different QMK maintainers (@drashna, @waffle87, @fauxpark) being assigned to this single pull request requesting various changes one after another, is time-consuming and inefficient. This has led to delays and halts in the merging process of my pull request, as I have to communicate with three different individuals to get it approved.
I understand that each QMK maintainer reviews pull requests and work on improving the QMK firmware in their free time. I appreciate their ongoing efforts to assist new QMK developers and keyboard designers. I also recognize the need for keyboard firmware files to adhere to specific formats, guidelines, and code optimization. As a developer, I hope that my pull request will be approved so that I can continue to contribute to and further develop this project.

drashna commented:
You mean like https://docs.qmk.fm/pr_checklist ? 

Also, while the document tries to be comprehensive, it's not *everything*.  And as for multiple people, there are different things that each of us are better about picking up on.

drashna commented on keyboards/protokeeb/rev1/rev1.c:
```diff
@@ -0,0 +1,28 @@
+#include "quantum.h"
+
+// Override the matrix read function to include the encoder button
+void matrix_read_cols_on_row(matrix_row_t current_matrix[], uint8_t current_row) {
```
You haven't tested this on hardware, correct?

atechofficials replied:
No, I have tested the firmware on the actual keyboard hardware, and the new encoder push button functionality code is working as intended. Both the rotary encoder and the encoder push button are functioning properly. When using VIA keymap, the user can successfully assign any keycodes to the encoder and its push-button using the [VIA configurator](https://usevia.app/)

zvecr replied:
Given this implementation, it cant be working correctly.

If you replace matrix_read_cols_on_row with one that does not read any of the other pins, then none of the actual matrix would get scanned....

atechofficials replied:
Let me check if all the keys of the switch matrix are working or not because last time when I changed the encoder push button functionality code according to @drashna's suggestions:

>You could add an additional column, and add a local matrix_read_cols_on_row that reads from the specific pin for the encoder at the column/row spot. It would allow the encoder to act like it was hooked into the switch matrix, normally. However, it would no longer be an ortho 4x4 layout. But it would allow for proper via support for the encoder, for instance.

I only tested the encoder and the encoder push button functionality, not the switch matrix functionality.

atechofficials replied:
@zvecr, you were right – the switch matrix keys are not working. I need to figure out a way for both the switch matrix scanning and encoder push button to work as intended. I'm open to any suggestions, as my coding skills are not that refined.

atechofficials replied:
I managed to figure out the code to implement encoder push button functionality without affecting the switch matrix scanning. I tested the firmware on the keyboard hardware and everything is working as intended. All the keys of the switch matrix are functioning properly, and the encoder, as well as the encoder push button, are also working as intended. Please review the code and approve this pull request.

rev1.c

```C
#include "quantum.h"

extern matrix_row_t matrix[]; // Declare the matrix array

void matrix_scan_kb(void) {
    // Read the encoder button state and update the matrix
    if (gpio_read_pin(ENCODER_BUTTON_PIN) == 0) {
        matrix[ENCODER_BUTTON_ROW] |= (1 << ENCODER_BUTTON_COL);
    } else {
        matrix[ENCODER_BUTTON_ROW] &= ~(1 << ENCODER_BUTTON_COL);
    }
}
```

atechofficials commented:
> You mean like https://docs.qmk.fm/pr_checklist ?
> 
> Also, while the document tries to be comprehensive, it's not _everything_. And as for multiple people, there are different things that each of us are better about picking up on.

I have already gone through the [PR Checklist](https://docs.qmk.fm/pr_checklist), [Contributing to QMK](https://docs.qmk.fm/contributing) and [Git Best Practices](https://docs.qmk.fm/newbs_git_best_practices) documentation before raising this pull request. 
Never mind, all the requested changes have been made and I see no point in further delaying this pull request. I appreciate @fauxpark's effort in taking the time to suggest using short key code names in keymaps. However, I prefer using long/full key code names, as they make it easier for me to understand, and this doesn't affect the functionality of the code.

fauxpark commented:
> However, I prefer using long/full key code names, as they make it easier for me to understand, and this doesn't affect the functionality of the code.

to clarify, some of them are deprecated. Please see the [keycode docs](https://docs.qmk.fm/keycodes).

atechofficials commented:
> > However, I prefer using long/full key code names, as they make it easier for me to understand, and this doesn't affect the functionality of the code.
> 
> to clarify, some of them are deprecated. Please see the [keycode docs](https://docs.qmk.fm/keycodes).

Hello @fauxpark,
I have replaced the deprecated keycodes in all the keymaps of the Protokeeb keyboard. Please review the changes and approve this pull request.

atechofficials commented:
Dear QMK Maintainers, @drashna, @waffle87, and @fauxpark,

I am writing to inform you that it has been 5 days since the requested changes for this keyboard have been completed. Please review and approve the changes and proceed with merging this pull request. 
Thank you.

fauxpark commented:
PRs are reviewed in our spare time. Please do not continually ping people about it.

atechofficials commented:
> PRs are reviewed in our spare time. Please do not continually ping people about it.

Thank you for letting me know. I apologize for the repeated pings and appreciate your team's efforts in reviewing pull requests in your spare time. I'll be patient and look forward to any feedback when you have the opportunity. Thanks again for your understanding and support.

atechofficials commented:
Hi QMK Maintainers, @drashna @waffle87 

I hope you're doing well. I wanted to follow up on my pull request, which has been pending for over a month. I've made all the requested changes and ensured that the code complies with the guidelines. If there’s anything further that needs to be addressed, please let me know. Otherwise, I would greatly appreciate it if you could review and merge the PR when you have a chance.

Thank you for your time and effort.

Best regards,
Mrinal

drashna reviewed (commented):
As of August 26, 2024, `qmk/qmk_firmware` is no longer accepting VIA-enabled keymaps as these have now transitioned to a repository under the VIA team's control.

As you've submitted a PR containing `via` or VIA-enabled keymap(s), this is your notice that they should be removed from this PR. You should now submit a secondary PR to the [VIA QMK Userspace](https://github.com/the-via/qmk_userspace_via) repository with your associated `via` or VIA-enabled keymaps instead.

atechofficials commented:
> As of August 26, 2024, `qmk/qmk_firmware` is no longer accepting VIA-enabled keymaps as these have now transitioned to a repository under the VIA team's control.
> 
> As you've submitted a PR containing `via` or VIA-enabled keymap(s), this is your notice that they should be removed from this PR. You should now submit a secondary PR to the [VIA QMK Userspace](https://github.com/the-via/qmk_userspace_via) repository with your associated `via` or VIA-enabled keymaps instead.

Hi QMK Maintainers,

Thank you for the clarification regarding the recent policy changes. I understand the need to transition VIA-enabled keymaps to the new repository. However, I would like to express my concern that my pull request was opened well before these changes were implemented.

I made all the requested changes promptly and ensured compliance with the guidelines at the time. It’s disheartening to see newer PRs being merged while mine remained pending for over a month, only to be affected by the new rules. Could you please help me understand if there was an issue with my PR before the policy change, or if there was another reason for the delay?

I genuinely want to contribute to the QMK community and improve my skills. Your guidance on how to best proceed in light of these changes would be appreciated.

Thank you again for your time and support.

Best regards,
Mrinal

fauxpark commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,90 @@
+            "knight": true,
+            "rainbow_mood": true,
+            "rainbow_swirl": true,
+            "rgb_test": true,
```
```suggestion
```

fauxpark commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,90 @@
+            "speed": 0,
+            "val": 50
+        },
+        "driver": "ws2812",
```
```suggestion
```

fauxpark commented on keyboards/protokeeb/rev1/keyboard.json:
```diff
@@ -0,0 +1,90 @@
+        },
+        "brightness_steps": 15,
+        "default": {
+            "animation": "static_light",
+            "hue": 0,
+            "on": true,
+            "sat": 255,
+            "speed": 0,
+            "val": 50
```
```suggestion
            "val": 50
```

atechofficials commented:
![QMK_FW_Compile_Success_1](https://github.com/user-attachments/assets/3713cf26-28ad-4118-8174-6cac3077b52d)
Hey there,

I wanted to bring up an issue with the Pull Request. It looks like the PR Lint Check failed after I made the changes to my keyboard.json file as suggested by @fauxpark. I ran the same test on my PC, and the lint check passed, and the firmware compiled successfully. I'm not sure why the QMK lint check failed on the pull request.

fauxpark commented:
You can see why above.

fauxpark commented:
So do you understand what the next step is?

atechofficials commented:
> So do you understand what the next step is?

Yes, I will remove the VIA keymap from the keyboard directory and create a separate pull request to submit the VIA keymap on the VIA QMK userspace repository.

atechofficials commented:
@drashna @fauxpark @waffle87 
I have removed the "via" keymap directory as you requested, and the keyboard firmware compiled successfully on my end. I would appreciate it if you could now proceed with merging this pull request. Alternatively, do I need to close this PR and open a new one?

waffle87 commented on keyboards/protokeeb/keymaps/default/rules.mk:
```diff
@@ -0,0 +1,2 @@
+ENCODER_MAP_ENABLE = yes
+TRI_LAYER_ENABLE = yes
```
```suggestion
```

atechofficials replied:
My default keymap utilizes tri-layer functionality. I believe that without this line of code in the rules.mk file, the tri-layer feature will be disabled.
https://docs.qmk.fm/features/tri_layer#tri-layers

zvecr replied:
Nope, that rules.mk controls the behaviour of the tri-layer keycodes, not the existence of the `update_tri_layer_state` function.

atechofficials replied:
So, from what I can understand is that the function update_tri_layer_state() does not depend on TRI_LAYER_ENABLE being set in rules.mk file and It is part of QMK's core layer handling and works regardless of whether TRI_LAYER_ENABLE is enabled.

waffle87 replied:
Correct.

atechofficials replied:
Ok if that is the case I no longer need the TRI_LAYER_ENABLE build option in the rules.mk file.

waffle87 commented on keyboards/protokeeb/rev1/config.h:
```diff
@@ -0,0 +1,50 @@
+#define WS2812_PIO_USE_PIO1
+
+/* RGB lighting effects and animations settings */
+#define RGBLIGHT_EFFECT_BREATHE_MAX 127
+#define RGBLIGHT_EFFECT_CHRISTMAS_INTERVAL 40
+#define RGBLIGHT_EFFECT_CHRISTMAS_STEP 2
+#define RGBLIGHT_EFFECT_KNIGHT_LED_NUM 4
+#define RGBLIGHT_EFFECT_KNIGHT_LENGTH 3
+#define RGBLIGHT_EFFECT_KNIGHT_OFFSET 0
+#define RGBLIGHT_EFFECT_SWIRL_RANGE 255
+#define RGBLIGHT_EFFECT_SNAKE_LENGTH 4
+#define RGBLIGHT_EFFECT_TWINKLE_LIFE 200
+#define RGBLIGHT_EFFECT_TWINKLE_PROBABILITY (1/127)
```
```suggestion
#define RGBLIGHT_EFFECT_BREATHE_MAX 127
```

atechofficials replied:
I understand your request to remove these lines, as most of these settings have default values. This avoids redefining the Effect and Animation Settings in my firmware configuration file.

waffle87 commented on keyboards/protokeeb/rev1/config.h:
```diff
@@ -0,0 +1,50 @@
+/* Bootloader timeout window in milli-seconds */
+#define RP2040_BOOTLOADER_DOUBLE_TAP_RESET_TIMEOUT 200U
+
+/* LED to blink when entering bootloader mode */
+#define RP2040_BOOTLOADER_DOUBLE_TAP_RESET_LED GP25
+
```
There is no LED connected to this pin in this keyboard's schematic. I advise against copy/pasting code.
![image](https://github.com/user-attachments/assets/8b21aeca-320b-4cb9-b3a3-44befc24cc84)
```suggestion
```

atechofficials replied:
If you have closely examined the PCB files for this keyboard, you would not accuse me of blindly copying and pasting code from other keyboard firmware developers. I am using a Raspberry Pi Pico (RP2040) development board as the main controller for this keyboard, and the built-in LED of the Raspberry Pi Pico is connected to GPIO25. Therefore, if I am using this built-in LED to visually indicate when the keyboard enters in bootloader mode, what is wrong with that?

waffle87 replied:
Ah, fair enough. Wonder if it would be wise to add a small comment like
```c
// Pi Pico's on board LED
#define RP2040_BOOTLOADER_DOUBLE_TAP_RESET_LED GP25
```
So there is not confusion for others using a different development board.

atechofficials replied:
Currently, the PCB of the Protokeeb Rev1 keyboard is compatible with the Raspberry Pi Pico (RP2040) and the Waveshare RP2040-Plus development boards. Any other RP2040 development boards will not be compatible unless they have the same footprint and pin assignments as the Raspberry Pi Pico (RP2040). Additionally, I will make sure to include that comment you suggested.

waffle87 commented on keyboards/protokeeb/rev1/mcuconf.h:
```diff
@@ -0,0 +1,26 @@
+
+/* RP2040 I2C Driver Selection */
+#undef RP_I2C_USE_I2C0
+#define RP_I2C_USE_I2C0 TRUE
```
What is this needed for? This keyboard does not have an I2C peripherals.

atechofficials replied:
I previously left these lines of code while testing an I2C OLED display module with the keyboard. I will remove them now that the I2C interface is no longer necessary.

waffle87 commented on keyboards/protokeeb/rev1/config.h:
```diff
@@ -0,0 +1,50 @@
+#define ENCODER_BUTTON_ROW 0
+#define ENCODER_BUTTON_COL 4
+
+/* Peripheral used */
+#define WS2812_PIO_USE_PIO1
+
```
```suggestion
```

atechofficials replied:
If I am not wrong the PIO driver is used by RP2040 to control the WS2812 RGB LEDs as mentioned here:
https://docs.qmk.fm/drivers/ws2812#pio-driver
https://docs.qmk.fm/drivers/ws2812#arm-pio-driver
Additionally, this line of code instructs the firmware compiler to use the PIO1 peripheral instead of PIO0 for ARM-based MCUs like RP2040.

zvecr replied:
We are aware of what the define does. The question is "I need to use PIO1 instead, because....?". The answer to that is you dont as you have nothing else that depends on PIO so its fine to use the defaults and the line needs to be removed.

atechofficials replied:
I understand now. These details need to be clarified in the QMK documentation, specifically in the WS2812 PIO driver section. When I first created the config files as a beginner in QMK, I assumed that PIO0 was being used by another function of the RP2040 MCU. So I thought that’s why there is an option to use PIO1 but I was wrong.

zvecr commented on keyboards/protokeeb/info.json:
```diff
@@ -0,0 +1,11 @@
+    "manufacturer": "A-Tech Officials",
+    "maintainer": "atechofficials",
+    "url": "https://github.com/atechofficials/protokeeb",
+    "bootloader_instructions": "Press and hold the FN Key (row-4, column-1) then press the key mapped to `QK_BOOT` i.e ENTER Key (row-4, column-4) to enter the bootloader mode.",
```
`bootloader_instructions` should only be hardware related, and not instructions on bootmagic or the reset keycode. 
```suggestion
```

atechofficials replied:
Ok understood.

atechofficials commented on keyboards/protokeeb/rev1/rev1.c:
```diff
@@ -0,0 +1,28 @@
+    } else {
+        matrix[ENCODER_BUTTON_ROW] &= ~(1 << ENCODER_BUTTON_COL);
+    }
+}
```
Done.

atechofficials commented:
Hi @zvecr, @waffle87, and @fauxpark,

I have implemented all the requested changes based on your feedback. Could you kindly re-review and update your reviews? This PR has been open for 7 months, and I’ve addressed all suggestions and modifications promptly.

The PR still shows a "merging is blocked" message, and I want to ensure there’s nothing else pending from my side. Please let me know if any further changes are needed.

Thank you for your time and assistance!

tzarc commented:
Please stop pinging maintainers. We understand that your pull request is important to you, but it is one of several hundred the team has to deal with, and whilst quoting seven months (actually just over five) a large portion of that time has been waiting for you to perform the requested changes.

The team will get to this in due course, asking for updates will not speed that up.

atechofficials commented:
> Please stop pinging maintainers. We understand that your pull request is important to you, but it is one of several hundred the team has to deal with, and whilst quoting seven months (actually just over five) a large portion of that time has been waiting for you to perform the requested changes.
> 
> The team will get to this in due course, asking for updates will not speed that up.

Hi @tzarc,

Thank you for your clarification. I understand that the maintainers are managing numerous pull requests, and I appreciate the effort that goes into maintaining the QMK repository. However, I would like to know if there is any intention to merge this pull request, as it has been open for several months.

If my contribution does not align with the project's priorities or standards, please let me know. I would prefer to have clarity rather than continue making minor refinements without a clear path forward. This would help prevent wasting both my time and the resources of the QMK team.

Thank you again for your feedback, and I look forward to hearing your thoughts.

fauxpark commented on keyboards/protokeeb/rev1/mcuconf.h:
This file can be deleted.

atechofficials replied:
Ok, Done.
